### ISSUE TITLE
Class for Paramters

### ISSUE THREAD

**Issue by jhauschild** [tenpy/tenpy#49]
Currently, we store parameters as dictionary.
Make a subclass of the python build-in dictonary to replace the funcitonality of the functions in 
`tenpy.tools.params`:

- initiliaze from a dictionary (that was given to an algorithm, make a copy) and a `name` argument to be used for `get`. Set `verbose` as attribute.
- method for checking verbosity
- overwrite the `get` with functionality similar to `get_parameter`, taking only the key and default
- check for unused parameters by defining a `__del__` method.

**Comment by jhauschild** [comment]
Maybe we should also think about how we document the individual parameters.
Currently, we have these huge tables. Maybe, it would be better to put them into a 
`Parameters` list?

**Comment by Lschoon** [comment]
I'm happy to implement this (and have already started playing with it a little). Collecting some questions here:

- Do you definitely want this to be a subclass of the built-in python dictionary? It seems to me that all the desired functionality can be obtained by keeping the initial dictionary as an attribute of some generic class. There are many online sources warning me against subclassing `dict`. It also seems that there is some functionality in the built-in dict that we might not want (mainly accessing a value through `d['key']` without supplying a default). What are the expected benefits of subclassing `dict`? Perhaps a `collections.abc.MutableMapping` (see [documentation](https://docs.python.org/3/library/collections.abc.html)) is the solution.
- Should the `verbose` attribute be pulled from the initial dictionary (my preference), or be set through `__init__()`?
- Do you want to keep the `asarray` option for `get_parameter` somehow?
- I don't fully understand your comment. Yes, the parameter documentation is fairly bulky, but given the complexity of our algorithms that is not so surprising. How would you expect this `Parameters` list to be formatted?

**Comment by Lschoon** [comment]
For context with my first question, the following code should implement most/all of the desired functionality, without any inheritance:
```python
class ParameterDict:
    def __init__(self, params, name):
        self.params = params
        self.verbose = params.get('verbose', 0)
        self.name = name
        self.unused = set(params.keys())

    def get_parameter(self, key, default):
        val = self.params.setdefault(key, default)  # get the value; set default if not existent
        self.print_if_verbose(key, default)
        self.unused.discard(key)  # (does nothing if already present)
        # if asarray:
        #     val = np.asarray(val)
        return val

    def print_if_verbose(self, key, default):
        name = self.name
        use_default = key not in self.params
        new_key = key in self.unused
        if self.verbose >= 100 or (new_key and verbose >= (2. if use_default else 1.)):
            defaultstring = "(default) " if use_default else ""
            print("parameter {key!r}={val!r} {defaultstring}for {name!s}".format(
                name=name, key=key, val=val, defaultstring=defaultstring))

    def unused_parameters(self, warn=None):
        unused = self.unused
        name = self.name
        if warn is not None:
            if len(unused) > 0:
                if len(unused) > 1:
                    msg = "unused parameters for {name!s}:\n{keys!s}"
                else:
                    msg = "unused parameter {keys!s} for {name!s}\n"
                warnings.warn(msg.format(keys=sorted(unused), name=warn))
        return unused
```

**Comment by jhauschild** [comment]
The use case of the class (let's call it simply `Parameters`) in tenpy is as follows:
- In the `__init__(self, params, ...)` of a Model/Engine/... class, we do a 
  `self.params = Parameters(params, self.__class__.__name__)`.
- Whenever we need to read out a parameter in a class method, we use
  `self.params.get(key, default)` if we have a default value, or 
  `self.params[key]` if it is a mandatory parameter without default value.
- The dict-like interface makes it easy to understand even if you don't know about the fact that `self.params` is a `Parameters` instance - it just behaves like dict in the sense of duct-typing. 
 (At the moment, people defining a new model need to understand what the `get_parameter` function is. It would be good if we can avoid that.)
- In the __del__ method of `Parameters`, we should check for unused parameters - in that way we don't need to write a `__del__` method for each class using it.
- Of course, we can also do the same for a long function like `DMRG.run()`.
- Pretty-printing could be quite useful.
- The class should be `Hdf5Exportable`.
- Methods for loading/writing a yaml file could be helpful - I found that quite useful in an ongoing project.

Now to answer your questions:
1. After reading on the topic of subclassing `dict`, I agree that it's probably a bad idea. I think you're right and we should inherit from `MutableMapping`. All of the `__getitem__`, `__setitem__` and `__delitem__` should probably be verbose.
2. From the dictionary.
3. No, we probalby don't need it. Using `np.assary(self.params.get(...))` is likely more readable than `self.params.get(..., asarray=True)`. To be honest, I'm not sure anymore, if it was a good idea to introduce it in the beginning.
4. Formatting the tables in reStructured Text is very annoying. In the model classes, we "cheat" away the tables by putting a "All parameters are collected in a single dict" before the `Parameters` section in the doc-string. We could do something similar in the DMRG class.
That leaves the problem, however, that we use the parameters in other methods than where we document them. Where do we consistently document the parameters? The straightforward answer is "In the class which converted the parameters". But that won't work for models - we need to document the standard model parameters inside the model class, but we also have the default arguments like `bc_MPS`, `lattice` etc. How to document that?
There came two solutions to my mind:
- Put the default values in a separate file (e.g. a yaml file) and document them there.
  Load them during `Parameters.__init__`. Would allow to check for misspelled parameter names already during that phase.
  Makes adding/modifying parameters very clumsy, though - one *has to* modify the default parameters file during coding. Where to put those files? How to handle user-defined classes like the models?
- Add a method like `Parameters.document(self, key, type_description, help_text)` to the `Parameters` class, and document the parameter right in the code where you use them.
Further, add a method `Parameters.help` to pretty-print/return that help text for a given or all the parameters. When building the documentation with Sphinx, we can initialize an instance of the model/engine/... with the default parameters and just include the help text (even including the default parameters!).
The advantage of this method would be that it allows to get all the parameters into each class documentation; the obvious disadvantage is that it gives a slight runtime overhead. However, we also get a runtime overhead simply be replacing the parameter dict with the `Parameters` class, and in most cases it's just a single function call in the beginning of a very long-running program. After all, we're writing Python code...

What do you think of those two suggestions?

**Comment by Lschoon** [comment]
I see. I will bee subclassing `MutableMapping`, then. Hopefully I will have some time today, but in any case this should be done before the 0.6.0 release.

I much prefer your second suggestion: I don't like having the default values in a separate file from the code that uses them. I also like how using the `document()` method makes it easier to sligthly change things later on.

We will indeed incur some overhead due to the repeated calls to `Parameters.document()` but I am not too worried about that. It will probably pay off to have all these calls in a separate method for each class, to avoid their `__init__()` methods being overcrowded with documentation.

**Comment by Lschoon** [comment]
A to-do list:

- [ ] Create a class `Parameters` as subclass of `MutableMapping`. Overwrite (at least) `__getitem__`, `__setitem__`,  `__delitem__` and `__del__`.
- [ ] Make `Hdf5Exportable`.
- [ ] Add read/write methods for yaml-files.
- [ ] Implement verbosity checks for getters and setters.
- [ ] Add documentation method and a `Parameters.help()` method.
- [ ] Add pretty-printing
- [ ] Have the Sphinx build include the help text. @jhauschild, I have never directly worked with Sphinx so am not quite sure how to do this.

I am not sure what you meant by your comment on long functions like `dmrg.run()`.

**Comment by Lschoon** [comment]
A couple more questions:

1. Do `save_hdf5()` and `from_hdf5()` need to be overwritten?
2. Do we want verbose behaviour for `__setitem__()` and `__delitem__()`? Do we want this verbose behaviour to be different from that in `__getitem__()`?
3. Is there a style guide for output generated by TeNPy somewhere? The output right now works, but looks a little rough.

**Comment by jhauschild** [comment]
1. They don't have to be overwritten if you just derive from `HDF5Exportable`;
  for "simple" classes with attributes which are only simple data types like lists/dicts/... it will just work out of the box.
  Overwriting these methods means to explicitly define the data structure in the exported files, and to drop duplicated/cached data. In that case, you don't even have to derive from `Hdf5Exportable`, the export/import uses duct-typing for user-defined classes.
  If one later on changes the internal data structure of the class (e.g rename/add attributes), these methods allow to keep the previous file format and just handle the new/renamed attributes. For example, I just provided the default value `False` for `explict_plus_hc` in the MPO.
2. Yes. No, apart from a slightly different message, probably.
3. No, there is no style guide. I'm open to suggestions!

**Comment by Lschoon** [comment]
Basic functionality is now in place. Some to-do's above are left open, they seem either non-essential or backwards compatible, and/or I don't know how to implement them.

Next up:
- [ ] Change algorithms to read in their parameters as `Parameters` object.
- [ ] Change parameter documentation to use `Parameters.document()`.
- [ ] Change `test_params.py` to use `Parameters` class.

**Comment by Lschoon** [comment]
I have edited all models, as well as `DMRG.py`, `TEBD.py`, `TDVP.py` and `MPS_Sweep.py` to use the new `Parameters`. Documentation is still to-do, as well as getting things like mixer and Lanczos parameters to use `Parameters`.

To avoid a problem with `tools.misc.any_nonzero()` accidentally setting the value of a parameter to `None`, I introduced a very clunky fix in https://github.com/tenpy/tenpy/commit/3384c695df11afa7cb8ac01528545606672c3dbf. This is ridiculously ugly and needs a better solution, so I will come back to this. I am thinking of having `Parameters.get()` check whether `default is None`, and stop it from setting `None` as parameter value, but there might be use cases where one would _want_  to set `None`. Will think about this.

**Comment by Lschoon** [comment]
Additional note: I just noticed that GitHub's automatic testing of my commits fails because it cannot `import yaml`. On my machine, all pytest tests complete without errors. @jhauschild, can we get GitHub to do a `pip install yaml` before running tests?

**Comment by jhauschild** [comment]
> Hopefully I will have some time today, but in any case this should be done before the 0.6.0 release.

To make that clear: I don't consider the anyways somewhat arbitrary deadline of the 0.6.0 release as strict, I just want to realease the next version soonish, when we have it ready.

> Have the Sphinx build include the help text. @jhauschild, I have never directly worked with Sphinx so am not quite sure how to do this.

I will have to look into that as well, but I'll do that (hopefully by the end of the week). As I said, I have at least a rough idea how it should work (and I'm quite confident that it will).

> To avoid a problem with `tools.misc.any_nonzero()` accidentally setting the value of a parameter to `None`, I introduced a very clunky fix in [3384c69](https://github.com/tenpy/tenpy/commit/3384c695df11afa7cb8ac01528545606672c3dbf). This is ridiculously ugly and needs a better solution, so I will come back to this. I am thinking of having `Parameters.get()` check whether `default is None`, and stop it from setting `None` as parameter value, but there might be use cases where one would _want_ to set `None`. Will think about this.

I think an elegant solution is to make `any_nonzero()` a method of the `Parameters` class, now that you implemented it. Inside the model class, where this function get's called, we can safely assume that the given parameters are an `Parameters` instance - we need to do that anyways, if we want to use the `document` method.

By the way, I noticed that the `print_if_verbose` doesn't work as expected, it will never print `(default)`. Can you fix that?

**Comment by Lschoon** [comment]
Good to know the deadline isn't strict -- we would have missed it by now! I have fixed both `any_nonzero()` and `print_if_verbose()`now.

**Comment by jhauschild** [comment]
I'm reading up on sphinx right now.
Before you actually implement the use of `parameters.document()` for the parameters, let me think a bit more about whether this is a good idea way to do it.

**Comment by jhauschild** [comment]
A quick update (I'm developing on the `sphinx-parameters` branch).

I realized that documenting the parameters dynamically doesn't help; it just shifts the problem of how to collect it. If we want to include the possible parameters into the documentation (and we definitely should!), we still need a python function at some point, which you can call to get all the parameters. 
If we do it dynamically, that function would need to do full DMRG/... runs; then building the Documentation would depend on the different algorithms to run, and we would need functions to setup those runs. Ugly!

The proper way to do it should be to document the parameters in the Docstrings - that's what docstrings are for.

Over the weekend, I've been digging into the sphinx code and how to write a sphinx extension; this turned out to be much harder than I anticipated.
But I still believe that sphinx is the proper way to handle this.

I'm aiming to implement the following (names might still change. Also, I'm not sure if I will manage to get all of it working, but from what I understood now, it should work....):

- [x] Implement a directive `.. parameterclass` to document an instance of the `Parameters` class.
	Sphinx can essentially handle it like a class definition, allowing to get a unique identifier, e.g. `tenpy.models.model.CouplingMPOModel.params`,
	Looks essentially like this (inside the doc-string of the `CouplingMPOModel`:
	```
	.. parameterclass :: param

		:entry first: Document parameter `first`.
		:type first: int
		:entry second: Document parameter `second`.
		:type second: float
	```

- [ ] Implement a directive `.. parameterentries` to define parameter entries of that instance in any docstring outside.
    Allow this to be called from any point of the documentation, and provide a back reference:
	```
	.. parameterentries: tenpy.models.model.CouplingMPOModel.params

		:entry third: Document parameter `third`.
		:type third: int
		:entry fourth: Document parameter `fourth`.
		:type third: str
	```
- [ ] During the sphinx build process, collect the parameter entry definitions of a given `Parameters` instance, and include them in the output of the `.. parameterclass` directive. Ideally, include back-links to where they were defined by ` ..parameterentries``!
- [ ] Make an option for the `.. parameterclass::` to include parameters from other instances; appending them to the output (marked appropriately).
	This will allow subclasses like `SpinChain` to define a own `.. parameterclass: params` and still list the parameters of the `CouplingMPOModel`.
- [ ] Finally, to make the docstrings nicer, it would be great if we can parse
   parameter entry definitions of the form
	```
	Used Paremeter-class Entries
	----------------------------
	:orig: tenpy
	first : int
		Document parameter `first`.
	second : float
		Document parameter `second`.
	```
	in the docstrings to get something like
	```
	.. parameterentries: tenpy.models.CouplingMPOModel

		:entry first: Document parameter `first`.
		:type first: int
		:entry second: Document parameter `second`.
		:type second: float
	```
	This is exactly what the `numpydoc` and `napoleon` sphinx extensions do, but it will require a bit hacking to use them for our purpose.
- [x] We can allow for references to a Parameters instance with :parcl:`~tenpy.models.model.CouplingMPOModel.params`.
- [ ] Moreover, we should allow reference to individual entries like :parentry:`~tenpy.model.model.CouplingMPOModel.params.first`.
	This will only work for parameter keys which are valid python variable identifies.

Does this appear reasonable?

**Comment by Lschoon** [comment]
I think your proposals sound reasonable, but have never worked with sphinx so can't tell for sure. I have a few questions/comments:

>  it would be great if we can parse
parameter entry definitions of the form

 ```
Used Paremeter-class Entries
 ----------------------------
 :orig: tenpy
 first : int
 	Document parameter `first`.
 second : float
 	Document parameter `second`.
```

> in the docstrings to get something like

```
 .. parameterentries: tenpy.models.CouplingMPOModel

 	:entry first: Document parameter `first`.
 	:type first: int
 	:entry second: Document parameter `second`.
 	:type second: float
```
Would we write the docstrings in the first or second form? What is the purpose of parsing to the second form?

> This will only work for parameter keys which are valid python variable identifies.

Does this mean we need to ensure parameters are read out into variables with the identifier equal to the parameter key, i.e., `var = params['var']` works, but `var = params['variable']` does not? That might be too restrictive, but could still be useful if we document it well.

Finally, I should probably read up on sphinx. Have you found the Sphinx documentation to be useful, or would you suggest other resources?

**Comment by jhauschild** [comment]
After a working on this for a few weekends and long nights, I've finished developing a sphinx extension: https://github.com/jhauschild/sphinx_cfg_options
That was way harder than I initially imagined, mostly because `docutils` is not so well documented, so I ended up reading a log of sphinx's source code to understand that part. 

Anyways... 
@Lschoon Take a look at the readme of that repo, and how it looks like in the built documentation, and let me know what you think.

I ended up chosing the name "config" for what we call the `Parameters` class here (and I strongly suggest renaming the `Parameters` class to `Config`), and I called the entries of the "config" "options". I think that makes it way clearer to distinguish it from other function "parameters".

The names for the `config` itself can be any identifiers, so we could choose to use `tenpy.algorithms.dmrg.TwoSiteDMRG.config`, but I'm very much in favor of just using `DMRG` for brevity. We might also use something like `self.config` and expand the `self` to the python class where it is documented.

The extension supports putting default values in the documentation, but we don't have to do that, I'm not sure if it is a good idea.

By the way, during developing the sphinx extension, I learned a bit about pythons (logging)[https://docs.python.org/3/howto/logging.html#logging-basic-tutorial], maybe we should start using that as well.

I'll include the extension into this branch shortly, and start updating the documentation accordingly.

**Comment by jhauschild** [comment]
to also anser your questions above: The second form is what sphinx uses by default for so-called fields, for example the parameters of a function. Initially, I wanted to use that for the options of the config, but you can't properly add cross-references for that, and it involved a lot of hacking around in the sphinx-code, so I decided against it later on.

I ended up writing the `.. cfg:option::` directive for parameters, and by default parse the non-indented content lines of the form `optname : opttype = optvalue` to something like
```
.. cfg:option:: optname
   :type: opttype
   :default: optvalue

```
such that 
```
.. cfg:config:: Vehicle

   max_speed : float = 220.
         Maximum speed of the vehicle in km/h.
   fuel : str = "gasoline"
      The type of fuel
```
looks to sphinx like
```
.. cfg:config:: Vehicle

   .. cfg:option:: max_speed
         :type: float
         :default:  220

         Maximum speed of the vehicle in km/h.

   .. cfg:option:: fuel
      :type: str
      :default:  "gasoline"

      The type of fuel

```
which in turn produces the results I wanted.
A bit hacky, but actually supported by sphinx, and the sphinx extensions numpydoc/napoleon do a similiar thing when handling the doc-strings - that's where I got the idea.

**Comment by Lschoon** [comment]
I like it! I have a few comments:

- Firstly, what is not completely clear to me is how to actually use this within TeNPy, or any actual Python code for that matter. A fully worked-out example would probably help. This is how I currently understand the usage:

```
class SomeClass():
    '''docstring

    .. cfg:config:: SomeClass
    
        some_option : float = 0
            This option is present for all SomeClass instances.
    '''

    def some_method():
        ''''docstring

        .. cfg:option :: other_option
            :config: SomeClass
            :type: bool
            :default: True
        '''
```
i.e., "option" descriptions get added in the methods that actually use those options and automatically parsed by Sphinx to appear at the `SomeClass` description. Is this correct? Also, is there a way to define multiple options for an already-existing "config" without having to repeat the lines
```
 .. cfg:option :: other_option
            :config: SomeClass
```
?

- The following code, from `README.rst` at https://github.com/jhauschild/sphinx_cfg_options results in "fuel" appearing _three times_ in the output at https://sphinx-cfg-options.readthedocs.io/en/latest/README.html. Two of those appearances contain the same information. I think that ideally, when the entire config documentation is reproduced if an option is added, we should not create extra output describing only that option. (Let me know if this makes sense, I can probably explain better with some screenshots.

```
.. cfg:config:: ElectricCar
   :include: Car

   fuel :
      Additional choice ``"battery"`` on top of what :cfg:option:`Vehicle.fuel` defines.
   hybrid : bool = False
      Whether the car has both an internal combustion engine and an electric motor, or not.
```

- I agree that the logging functionality looks interesting. This could replace all the output we currently print to the console. Let's create an "Enhancement" issue for this, as it would require overhauling most of the output functionality.

- I will rename the `Parameter` class as `Config`.

**Comment by jhauschild** [comment]
> A fully worked-out example would probably help.

Did you see https://sphinx-cfg-options.readthedocs.io/en/latest/mylib.html ?


> i.e., "option" descriptions get added in the methods that actually use those options and automatically parsed by Sphinx to appear at the `SomeClass` description. Is this correct? Also, is there a way to define multiple options for an already-existing "config" without having to repeat the lines [...]?

Instead of adding the options individually, you can define another `config`; they don't need to be unique. Moreover, you can supress including the `list` in the beginning with `:nolist:`; so in the `some_method`, you can do the following:
```
.. cfg:config:: SomeClass
    :nolist:
    
    extra_option1 : type = value
         Description of `ext_option1`.
    extra_option2 : type = value
         Description of `ext_option2`.
```

> The following code, from `README.rst` at https://github.com/jhauschild/sphinx_cfg_options results in "fuel" appearing three times in the output at https://sphinx-cfg-options.readthedocs.io/en/latest/README.html. Two of those appearances contain the same information. I think that ideally, when the entire config documentation is reproduced if an option is added, we should not create extra output describing only that option. (Let me know if this makes sense, I can probably explain better with some screenshots.

It appears once [as option](https://sphinx-cfg-options.readthedocs.io/en/latest/README.html#cfg-option-ElectricCar.fuel) and twice in tables, [here](https://sphinx-cfg-options.readthedocs.io/en/latest/README.html#cfg-config-ElectricCar) and [here](https://sphinx-cfg-options.readthedocs.io/en/latest/README.html#cfg-config-0).

The option definition (first link) is what `:cfg:option:`ElectricCar.fuel` points to, and is the only place were the *full* documentation of the option is included, and parsed.

The tables just include the *first line* summary of the option 
You can see that in the option `max_speed` of the very first example.
Sure, if the option is just one line long, as in that case, that's all the information, but it might as well not be.
This behaviour is similar as if you define a method with a doc-string in a class: the first line of the docstring appears in the method-summary-list of the class definition, but the full doc-string is only included once.
Also, the tables right now don't parse the summary; I couldn't figure out how to do that without breaking links and possibly the table layout.
Inside the tables, you can click on the name to get to the actual definition - which might be somewhere completely different, that's the whole point of this sphinx extension.

> I will rename the Parameter class as Config.

Thank you!


A few points, that I noticed during writing this answer:

- [ ] I should probably define a short-hand `.. cfg:configoptions:: SomeClass` that automatically sets the `:nolist:` option. 
- [ ] Also, if `:nolist:` is used, it should provide a link to the `config` definition with the table.
- [ ] We should add something like `[... more description]` in the tables if the summary does not contain the full definition of the option.
- [ ] In the TeNPy docs, we should also highlight the target of a link; that's a really usefull feature. I don't know why theread-the-docs theme doesn't do that...

**Comment by Lschoon** [comment]
> Did you see https://sphinx-cfg-options.readthedocs.io/en/latest/mylib.html ?

I did not. (Note for posterity: since renamed to https://sphinx-cfg-options.readthedocs.io/en/latest/example_lib.html)


> The tables just include the first line summary of the option

I hadn't realised this. That settles my concerns.

Is there anything left to do on this? I just merged the `master` branch so there are currently no merge conflicts.

**Comment by jhauschild** [comment]
> Is there anything left to do on this?

- [ ] There are still a lot of deprecation warnings when running the tests.
- [x] There was still a to-do for setdefault method
- [ ] Fix the to-do from the `any_nonzero`.
- [ ] Do you agree that it is a better idea to document the options in the doc-strings instead of dynamically with `Config.document(...)`? Then we should probably remove this feature (`Config.document() & `Config.documentation`) again.
- [ ] To be consistent with the documentation, it might make sense to rename the `Config.params` to `Config.options` and talk about "options" exclusively.
- [ ] Finally, we should update all the doc-strings where we use the options to use the new sphinx extension...

**Comment by Lschoon** [comment]
I am trying to update docstrings to use the sphinx extension. I have cloned the extension and included it in `tenpy/doc/conf.py` (both as a sys.path reference as well as in the list of extensions). However, a trial docstring does not get parsed when I call `make html`. 

I do get a warning: 

> WARNING: `cfg:option` 'chi_list' in /home/leon/tenpy/tenpy/algorithms/mps_sweeps.py:docstring of tenpy.algorithms.mps_sweeps.Sweep, line 21, **belongs to a non-indexed, unknown config** UNKNOWN (-> Typo?)

(emphasis mine)

A MWE of this docstring (note that neither way of including an option generates any html output):
```
.. cfg:config :: MPS_sweep

        .. cfg:option:: chi_list
            :type: dict or ``None``
            :default: ``None``

            Some description

        chi_list : dict | ``None``
            Some description 
```

Have I missed some setup option?

**Comment by jhauschild** [comment]
I had added the extension in 212308402d1cd2f7b26ce7f82b692a9a69e655c2, but didn't include it in the `extensions` of the `conf.py`. My guess is that the extension module was included twice for your config, which might have cause the problems. Sorry!
I've now added it to the `conf.py` it in c8761bdf6fb47296e0ef48e8c9f9d71c9399e338,
and checked that it works with documenting the `tenpy.algorithms.truncation.truncate` options as a `.. cfg:config`; this shows up in the built documentation as expected.

**Comment by Lschoon** [comment]
Can confirm this works. I hope to start moving around documentation later today.

**Comment by Lschoon** [comment]
2e92abbff8b0eb2d1207f17be2bb774192d9763d contains a first attempt at updating the docstrings in `mps_sweeps.py` with the new config options. A few questions arose from doing this:

1. I have made an effort to give documentation for a particular option in the docstring of the function where it is first read out (or in the class docstring if this is in `__init__()`. This seems the most reasonable (although if an option is read out twice in different places, it might lead to duplicate entries). We should probably note somewhere what the convention on this is.
2. Many of the options documented in  this way are read out into class attributes, sometimes with sligthly different names. This leads to redundant documentation. What should we do here?

**Comment by jhauschild** [comment]
Thanks!

bb6ea28757694efd1311fea45539f1eec988f936 defines some guidelines on documenting the config options.

1. Yes, we want to document the options where they get read out by the parameters.
2. The class attribute should still be documented (just to make clear that there is the attribute), but it can be as simple as for example
   ``` 
   chi_list : 
      See :cfg:option:`Sweep.chi_list`.
   ```

When generating the documentation, I noticed that the options documented in
`init_env` appear many times in the summary, because the `init_env` is documented in each subclass. Originally, I had intentionally allowed to include all the options if they were defined multiple times, but now this doesn't seem like a good idea any more...
I need to think about the best way to handle this. Maybe we shouldn't allow documenting a single option multiple times -- at least not with different content.
On the other hand, a subclass still needs to be able to overwrite what the option means, if it just uses a different definition (as for the fuel-expamle in the Electric car).
The best solution is probably to use inheritance rules as for class attributes/methods.

**Comment by jhauschild** [comment]
All options of a given name are only included once in the summary now.
Options included from other configs have a lower priority; but in case of ambiguity it's just the first one appearing in the documentation.

I've update the Lanczos and TEBD documentation to use the config directives.

**Comment by Lschoon** [comment]
I have updated DMRG documentation now. One hack (that I'm not very happy with) is that I included the `active_sites` option in the config for the `DMRGEngine` class, even though it is not used there. This option is used by `tenpy.algorithms.DMRG.run()` to decide which engine to run, but does get passed to it via the `Config` dictionary. Let me know if you have a better solution.

**Comment by Lschoon** [comment]
Model codes are now updated as well. One note resulting from that: we had a lot of model options collated on the same line, e.g.
```
Jxx, Jz, hz : float | array
            Couplings as defined for the Hamiltonian above.
```

This does not get parsed by the sphinx extension, so I have (for now) resorted to giving each parameter its own line. That might not be ideal, though (leads to eaven bulkier docstrings). Could the extension be changed so as to parse options like above?

**Comment by jhauschild** [comment]
Thanks for updating the configs.
As a side comment: unfortunately, the `.. cfg:config::` directive 
requires a full blank line separating the content and possible options, this is important for correct parsing.

Thanks for the suggestion of comma-separation options, this is now allowed (enabled by an option in conf.py).
I've update the models.

For the DMRG, the solution is to make a separate `config` inside the `run` function, which includes the `TwoSiteDMRG` and `SingleSiteDMRG` config.
The direction of includes should always be determined by how the `options` dictionary get's passed on:
If function/class `A` passes the `options` on to `B`, the config of `A` should include the config of `B`, because every option of B is then also availabe to A.
Note that this implies that configs of subclasses include the configs of the base classes - e.g., XXZChain includes the config of CouplingMPOModel.
Methods reading out from `self.options` don't get a separate config.

I've updated everything to follow these rules.

----------

Great, it seems that we can finally merge this branch into master :+1: 

One last thing: I noticed that the `config` documentation can get quite lengthy, 
and separate the function/class Parameters and Return section very far from the signature/docstring.
I therefore configured sphinx to allow a section 
```
Options
-------
```
in the docstrings, such that we can place the config options below the Parameters/Returns section.
This also allows to add general comments to it.
